clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -350,6 +353,92 @@ public synchronized void update(int requestVersion, MetadataResponse response, b
+            if (currentLeader.epoch.isPresent() && newLeader.epoch.get() <= currentLeader.epoch.get()) {
+                log.trace("For {}, incoming leader({}) is not-newer than the one in the existing metadata {}, so ignoring.", partition, newLeader, currentLeader);
+                continue;
+            }
+            if (!newNodes.containsKey(newLeader.leaderId.get())) {
+                log.trace("For {}, incoming leader({}), the corresponding node information for node-id {} is missing, so ignoring.", partition, newLeader, newLeader.leaderId.get());
+                continue;
+            }
+            if (!this.cache.partitionMetadata(partition).isPresent()) {
+                log.trace("For {}, incoming leader({}), no longer has cached metadata so ignoring.", partition, newLeader);
+                continue;
+            }
+
+            MetadataResponse.PartitionMetadata existingMetadata = this.cache.partitionMetadata(partition).get();
+            MetadataResponse.PartitionMetadata updatedMetadata = new MetadataResponse.PartitionMetadata(
+                existingMetadata.error,
+                partition,
+                newLeader.leaderId,
+                newLeader.epoch,
+                existingMetadata.replicaIds,
+                existingMetadata.inSyncReplicaIds,
+                existingMetadata.offlineReplicaIds
+            );
+            updatePartitionMetadata.add(updatedMetadata);
+
+            lastSeenLeaderEpochs.put(partition, newLeader.epoch.get());
+
+            updatedPartitions.add(partition);
+        }
+
+        if (updatedPartitions.isEmpty()) {
+            log.debug("No relevant metadata updates.");
+            return updatedPartitions;
+        }
+
+        // Get topic-ids for filtered partitions from existing topic-ids.
+        Map<String, Uuid> existingTopicIds = this.cache.topicIds();
+        Map<String, Uuid> filteredTopicIds = updatePartitionMetadata.stream()
+            .filter(e -> existingTopicIds.containsKey(e.topic()))
+            .collect(Collectors.toMap(e -> e.topic(), e -> existingTopicIds.get(e.topic())));
+
+        Set<String> updatedTopics = updatePartitionMetadata.stream().map(MetadataResponse.PartitionMetadata::topic).collect(Collectors.toSet());
+
+        if (log.isTraceEnabled()) {
+            updatePartitionMetadata.forEach(
+                partMetadata -> log.trace("For {} updating to leader-id {}, leader-epoch {}.", partMetadata, partMetadata.leaderId.get(), partMetadata.leaderEpoch.get())
+            );
+        }
+
+        this.cache = cache.mergeWith(cache.clusterResource().clusterId(), newNodes,
+                    updatePartitionMetadata, Collections.emptySet(), Collections.emptySet(),
+                    Collections.emptySet(), cache.cluster().controller(), filteredTopicIds,
+                    (topic, isInternal) -> !updatedTopics.contains(topic));
+        clusterResourceListeners.onUpdate(cache.clusterResource());
+
+        return updatedPartitions;
```

Comment:
Can we just pull the set of partitions out like so:

```suggestion
        return updatePartitionMetadata.stream()
            .map(MetadataResponse.PartitionMetadata::topicPartition)
            .collect(Collectors.toSet());
```